File: 2552-maximum-sum-of-distinct-subarrays-with-length-k/maximum-sum-of-distinct-subarrays-with-length-k.java
class Solution {
    public long maximumSubarraySum(int[] nums, int k) {
        int st = 0;
        long sum = 0, ans = 0;
        Map<Integer, Integer> freq = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            freq.put(nums[i], freq.getOrDefault(nums[i], 0) + 1);
            sum += nums[i];

            // Ensure all elements are unique
            while (freq.get(nums[i]) > 1) {
                freq.put(nums[st], freq.get(nums[st]) - 1);
                sum -= nums[st];
                if (freq.get(nums[st]) == 0) freq.remove(nums[st]);
                st++;
            }

            // If window size is exactly k, check for max sum
            if (i - st + 1 == k) {
                if (freq.size() == k) {
                    ans = Math.max(ans, sum);
                }

                // Remove leftmost element to slide the window
                freq.put(nums[st], freq.get(nums[st]) - 1);
                sum -= nums[st];
                if (freq.get(nums[st]) == 0) freq.remove(nums[st]);
                st++;
            }
        }

        return ans;
    }
}

File: 2552-maximum-sum-of-distinct-subarrays-with-length-k/maximum-sum-of-distinct-subarrays-with-length-k.cpp
class Solution {
public:
    long long maximumSubarraySum(vector<int>& nums, int k) {
        int st = 0;
        long long ans = 0;
        long long sum = 0;

        map<int, int> mpp;

        for(int i = 0; i < nums.size(); i++){
            mpp[nums[i]]++;
            sum += nums[i];

            if(mpp[nums[i]] >= 2){
                while(mpp[nums[i]] >= 2){
                    mpp[nums[st]]--;
                    sum -= nums[st];

                    if(mpp[nums[st]] == 0){
                        mpp.erase(nums[st]);
                    }
                    
                    st++;
                }
            }

            if(i - st + 1 == k){
                if(mpp.size() == k){
                    ans = max(ans, sum);
                    
                }

                mpp[nums[st]]--;
                sum -= nums[st];

                if(mpp[nums[st]] == 0){
                    mpp.erase(nums[st]);
                }
                
                st++;


            }
        }

        return ans;
    }
};